# Data Wrangling Orders and Products

# Importing libraries and Data sets

# Wrangling Orders

## Consistency Check on days_since_prior_order

### Cross Tab for days_since_prior_order

## Rename DOW column

# Creating a Data Dictionary

## Transpose Data Dictionary 

## Set Data Dictionary

# Exercise 4.4 Questions

## - Sales Questions

# Save Datasets

------------------

# Importing libraries and Data sets

In [1]:
#Importing Libraries
import pandas as pd
import numpy as np
import os

In [2]:
#creating DF pathway
path=(r'C:\Users\colmu\Desktop\Career Foundry\Achievement 4\Instacart Basket Analysis\Data')

In [3]:
#Import dataframe orders
df_ord= pd.read_csv(os.path.join(path,'original data','orders.csv'), index_col=False)

In [4]:
#import dataframe products
df_prod = pd.read_csv(os.path.join(path,'original data','product.csv'), index_col=False)

In [5]:
#import df department ID
df_dept = pd.read_csv(os.path.join(path,'original data','departments.csv'), index_col=False)

# Wrangling Orders Dataset

In [6]:
#Reading Task Drop orders.eval colmn
df_ord= df_ord.drop(columns = ['eval_set'])

## Consistency Check on days_since_prior_order

In [6]:
#Reading Task Orders value count days_since_prior_order
df_ord['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

### Cross Tab for days_since_prior_order

In [6]:
crosstab = pd.crosstab(df_ord['days_since_prior_order'], df_ord['order_number'], dropna = False)

In [7]:
crosstab.to_clipboard()

## Rename DOW column

In [7]:
#reading task rename dow colmn to days of week
df_ord.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

#reading task column check
df_ord.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [10]:
#reading task redefine order_id as string
df_ord['order_id'] = df_ord['order_id'].astype('str')

# Create Data Dictionary

## Transpose Data Dictionary 

In [6]:
df_dept.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [7]:
#Reading task tranposing DF_dep
df_dept.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [8]:
#reading task rename dept df
df_dept_t = df_dept.T

In [9]:
#reading task reset DF index dept
df_dept_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [10]:
#reading task create new header
new_header = df_dept_t.iloc[0]

In [11]:
#reading task create new  dpt_df without 1st row
df_dep_t_new = df_dept_t[1:]

In [12]:
#reading task attaching header to new dpt_df
df_dep_t_new.columns = new_header

In [13]:
df_dep_t_new.head()

department_id department
1                 frozen
2                  other
3                 bakery
4                produce
5                alcohol

## Set Data Dictionary

In [14]:
#reading task creating data dict
data_dict=df_dep_t_new.to_dict('index')

# Exercise Questions

In [21]:
#Task 2 Change different variable in the df_ords df to a different,more approepirate data type [user_id tp string]
df_ord['user_id']=df_ord['user_id'].astype('string')

In [22]:
#Task 2 checking user id to confirm data type has updated to string 
df_ord['user_id'].dtype

string[python]

In [23]:
#Task 3 Updating an unintuitive name in orders_DF [order_number to client_order_number]
df_ord.rename(columns={'order_number':'clients_order_number'}, inplace=True)

In [24]:
#Task 4 when is the busiest hour for placing orders [Answer 10 AM]
df_ord['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

In [25]:
#Task 5 what is the meaning of value 4 in the department_id column
print(data_dict.get('4'))

{'department': 'produce'}


## Sales Questions

In [26]:
#Task 6 Breakfast sales 
df_bfsales=df_prod[df_prod['department_id']==14]

In [27]:
#Task 6 Breakfast sales Observations
df_bfsales

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [28]:
# task 7 create a list of products for dinner parties for clients
df_prod_dinnerprtycont= df_prod.loc[df_prod['department_id'].isin([5,7,12,20])]

In [29]:
# Task 7 Observations
df_prod_dinnerprtycont

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

#task 8 How many rows does the last dataframe you created have? [answer 7650 rows]

### create profile for user 1

In [31]:
#Task 9 Profile on User_id 1
df_ord_user1=df_ord.loc[df_ord['user_id']=='1']

In [32]:
#Task 10 User_id 1 Behaviors [Includes all data from orders.csv]
df_ord_user1

order_id user_id  clients_order_number  orders_day_of_week  \
0   2539329       1                     1                   2   
1   2398795       1                     2                   3   
2    473747       1                     3                   3   
3   2254736       1                     4                   4   
4    431534       1                     5                   4   
5   3367565       1                     6                   2   
6    550135       1                     7                   1   
7   3108588       1                     8                   1   
8   2295261       1                     9                   1   
9   2550362       1                    10                   4   
10  1187899       1                    11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

# Save datasets

In [35]:
#Task 12 Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data”
df_ord.to_csv(os.path.join(path,'Prepared Data', 'orders_wrangled.csv'), index=False)

In [18]:
#Task 13 Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data”
df_dep_t_new.to_csv(os.path.join(path,'Prepared Data', 'departments_wrangled.csv'), index=False)

In [16]:
#Task 13 Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data”
df_dep_t_new.to_csv(os.path.join(path,'Prepared Data', 'departments_wrangled.csv'))